### DISCUSSION TITLE
When will IORING_CQE_F_MORE not be set while using multishot-accept?

### DISCUSSION THREAD

**Discussion by chimaoshu** [axboe/liburing#822] [Q&A] [answered]
~In `io_uring_prep_recv_multishot`, `IORING_CQE_F_MORE` will not be set in CQE flags when sever has read out of all data from a client `send()`.~(edit: that's not true) How about the `io_uring_prep_multishot_accept`? Under what circumstance will the listen_fd not "generate further CQEs"?

For example, if a client has `connect()` to the server, and multishot-accept successfully accepts it and generates a CQE, and there is no other client trying to `connect()` to the server at this time, will `IORING_CQE_F_MORE` be set?

According to [Linux kernel](https://github.com/torvalds/linux/blob/HEAD/io_uring/net.c#L1350), it seems that multishot-accept will always set `IORING_CQE_F_MORE` in CQE once the `accept` operation is successfully done, am I right?

**Comment by axboe** [comment, marked as answer]
It will remain set until an error has occurred accepting a connection. accept doesn't know if nobody else is trying to connect yet, that's a situation that could change at any point in time. Like any multishot request, it'll remain active until ones of these are true:

1) An error happens while accepting a connection
2) It's actively canceled by the application

**Comment by redbaron** [comment]
> In io_uring_prep_recv_multishot, IORING_CQE_F_MORE will not be set in CQE flags when sever has read out of all data from a client send()

Is it true? I thought recv remains active as long as connection stays up

**Reply by axboe** [reply]
No it's not true, where is that from? It'll stay active with the same rules as mentioned in https://github.com/axboe/liburing/discussions/822#discussioncomment-5281767

**Reply by axboe** [reply]
If it were true, there would be no difference between a normal recv and a multishot recv at that point. If all the data has been received, poll will be armed again and trigger when more data is available in the socket. Repeat until one of the termination conditions are met.

**Reply by chimaoshu** [reply]
> No it's not true, where is that from? It'll stay active with the same rules as mentioned in [#822 (comment)](https://github.com/axboe/liburing/discussions/822#discussioncomment-5281767)

So a single `io_uring_prep_recv_multishot()` can handle multiple `send()` from clients?
It may sound funny that I got that wrong info from Microsoft's AI new bing. I asked it about multishot-recv and it told me that multishot-recv is used in a single recv with large data size, also told me that IORING_CQE_F_MORE will not be set if server has read out of all data from a client `send()`.

**Reply by axboe** [reply]
Of course, otherwise it'd just be a normal send. Don't trust any of the LLM generative AIs, they make shit up all the time... So yes, both of the things it told you is utter garbage.

**Reply by axboe** [reply]
For the last one, it has nothing to do with IORING_CQE_F_MORE, but IORING_CQE_F_SOCK_NONEMPTY is set in the cqe if there socket was NOT empty when the receive was done. Again this really has nothing to do with multishot, that's a generic receive property.
